### PR TITLE
API Theme stacking

### PR DESCRIPTION
**Pull request by hafriedlander**
Not enough docs or tests yet. Raising for PR.

### CODE REVIEW

File: core/manifest/ThemeManifest.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
+<?php
+
+/**
+ * A class which builds a manifest of all themes (which is really just a directory called "templates")
+ *
+ * @package framework
+ * @subpackage manifest
+ */
+class ThemeManifest {
```
Comment (sminnee) [inline core/manifest/ThemeManifest.php]: Maybe as well introduce it within a namespace from the start? We did this, e.g., with new assets work.

Arguably this is more a part of "view" than "core"—if you were to remove the view you'd also remove this.

Comment (hafriedlander) [inline core/manifest/ThemeManifest.php]: Ah yup, good call. Will change.

File: docs/en/04_Changelogs/4.0.0.md [hunk 2]
```diff
@@ -393,6 +393,29 @@ all changed project files.
 This will resolve the majority of upgrading work, but for specific changes that will
 require manual intervention, please see the below upgrading notes.
 
+### Update theme setting
+
+SSViewer no longer takes a single theme - instead themes are set in a cascade. To upgrade from 3, 
+change:
+
+	theme: simple
```
Comment (sminnee) [inline docs/en/04_Changelogs/4.0.0.md]: Can this syntax be supported but deprecated?

File: forms/gridfield/GridFieldToolbarHeader.php [hunk 3]
```diff
@@ -13,7 +13,7 @@ class GridFieldToolbarHeader implements GridField_HTMLProvider {
 
 	public function getHTMLFragments( $gridField) {
 		return array(
-			'header' => $gridField->renderWith('GridFieldToolbarHeader')
+			'header' => $gridField->renderWith('Includes/GridFieldToolbarHeader')
```
Comment (sminnee) [inline forms/gridfield/GridFieldToolbarHeader.php]: Has the necessity of this kind of change been mentioned in the upgrade docs?

File: docs/en/04_Changelogs/4.0.0.md (hunk 2 above)
Comment (hafriedlander) [inline docs/en/04_Changelogs/4.0.0.md]: Actually, it is now. Fixed it after writing this. Will update. (The old behaviour supported it but didn't add `$default` so wasn't useful).

File: forms/htmleditor/TinyMCEConfig.php [hunk 4]
```diff
@@ -419,13 +419,22 @@ protected function getEditorCSS() {
 			Director::absoluteBaseURL(),
 			FRAMEWORK_ADMIN_DIR . '/client/dist/styles/editor.css'
 		);
-		if($theme = SSViewer::get_theme_folder()) {
-			$editorDir = $theme . '/css/editor.css';
+
+	    foreach(SSViewer::get_themes() as $theme) {
```
Comment (sminnee) [inline forms/htmleditor/TinyMCEConfig.php]: Err?

File: tests/core/manifest/TemplateLoaderTest.php [hunk 5]
```diff
@@ -26,6 +26,8 @@ public function setUp() {
 	 * Test that 'main' and 'Layout' templates are loaded from module
 	 */
 	public function testFindTemplatesInModule() {
```
Comment (sminnee) [inline tests/core/manifest/TemplateLoaderTest.php]: Should we delete these tests? Refactor them? Mark them as skipped or incomplete?

Comment (hafriedlander) [inline tests/core/manifest/TemplateLoaderTest.php]: Yeah, the whole file needs replacing with a different one that tests the new behaviour. Change to this PR with better docs & tests should be up today.

**Comment by sminnee** [comment]
Initial reading of the code looks reasonable, only minor suggestions relating to namespacing.

**Comment by hafriedlander** [comment]
CMS tests won't pass until this & https://github.com/silverstripe/silverstripe-cms/pull/1548 is merged.

**Comment by dhensby** [comment]
SiteConfig module needs a patch?

File: view/TemplateLoader.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+	public function __construct($base = null) {
+		$this->base = $base ? $base : BASE_PATH;
+	}
+
+	public function addSet($set, $manifest) {
+		$this->sets[$set] = $manifest;
+	}
+
+	public function getPath($identifier) {
```
Comment (tractorcow) [inline view/TemplateLoader.php]: Needs PHPDoc

**Pull request merged**